src/litdata/streaming/fs_provider.py

```diff
@@ -0,0 +1,262 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple
+from urllib import parse
+
+from litdata.constants import _GOOGLE_STORAGE_AVAILABLE
+from litdata.streaming.client import S3Client
+
+
+class FsProvider(ABC):
+    def __init__(self, storage_options: Optional[Dict[str, Any]] = {}):
+        self.storage_options = storage_options
+
+    @abstractmethod
+    def upload_file(self, local_path: str, remote_path: str) -> None:
+        raise NotImplementedError
+
+    def download_file(self, remote_path: str, local_path: str) -> None:
+        raise NotImplementedError
+
+    def download_directory(self, remote_path: str, local_directory_name: str) -> str:
+        raise NotImplementedError
+
+    def copy(self, remote_source: str, remote_destination: str) -> None:
+        raise NotImplementedError
+
+    def list_directory(self, path: str) -> List[str]:
+        raise NotImplementedError
+
+    def delete_file_or_directory(self, path: str) -> None:
+        raise NotImplementedError
+
+    def exists(self, path: str) -> bool:
+        raise NotImplementedError
+
+    def is_empty(self, path: str) -> bool:
+        raise NotImplementedError
+
+
+class GCPFsProvider(FsProvider):
```

Comment:
I don't see retries. We can merge without but we need to consider for a follow up PR.

Reply:
The retries are coming directly from the storage_options.